exchanges/poloniex/poloniex_wrapper.go

```diff
@@ -417,172 +522,365 @@ func (e *Exchange) GetWithdrawalsHistory(ctx context.Context, c currency.Code, _
 	}
-	var err error
-	pair, err = e.FormatExchangeCurrency(pair, assetType)
+	fPair, err := e.FormatExchangeCurrency(pair, assetType)
 	if err != nil {
 		return nil, err
 	}
 
 	var resp []trade.Data
-	ts := timestampStart
-allTrades:
-	for {
-		var tradeData []TradeHistory
-		tradeData, err = e.GetTradeHistory(ctx,
-			pair.String(),
-			ts.Unix(),
-			timestampEnd.Unix())
+	switch assetType {
+	case asset.Spot:
+		tradeData, err := e.GetTrades(ctx, fPair, 1000)
 		if err != nil {
 			return nil, err
 		}
-		for i := range tradeData {
-			var tt time.Time
-			tt, err = time.Parse(time.DateTime, tradeData[i].Date)
+		for _, td := range tradeData {
+			if td.CreateTime.Time().After(timestampEnd) ||
+				td.CreateTime.Time().Before(timestampStart) {
+				continue
+			}
+			side, err := order.StringToOrderSide(td.TakerSide)
 			if err != nil {
 				return nil, err
 			}
-			if (tt.Before(timestampStart) && !timestampStart.IsZero()) || (tt.After(timestampEnd) && !timestampEnd.IsZero()) {
-				break allTrades
-			}
-			var side order.Side
-			side, err = order.StringToOrderSide(tradeData[i].Type)
+			resp = append(resp, trade.Data{
+				Exchange:     e.Name,
+				CurrencyPair: fPair,
+				AssetType:    assetType,
+				Side:         side,
+				Price:        td.Price.Float64(),
+				Amount:       td.Amount.Float64(),
+				Timestamp:    td.CreateTime.Time(),
+			})
+		}
+	case asset.Futures:
+		tradeData, err := e.GetFuturesExecution(ctx, fPair.String(), 0)
+		if err != nil {
+			return nil, err
+		}
+		for _, fExecInfo := range tradeData {
+			side, err := order.StringToOrderSide(fExecInfo.Side)
 			if err != nil {
 				return nil, err
 			}
 			resp = append(resp, trade.Data{
 				Exchange:     e.Name,
-				TID:          tradeData[i].TradeID,
-				CurrencyPair: pair,
+				CurrencyPair: fPair,
 				AssetType:    assetType,
 				Side:         side,
-				Price:        tradeData[i].Rate,
-				Amount:       tradeData[i].Amount,
-				Timestamp:    tt,
+				Price:        fExecInfo.Price.Float64(),
+				Amount:       fExecInfo.Amount.Float64(),
+				Timestamp:    fExecInfo.CreationTime.Time(),
 			})
-			if i == len(tradeData)-1 {
-				if ts.Equal(tt) {
-					// reached end of trades to crawl
-					break allTrades
-				}
-				if timestampStart.IsZero() {
-					break allTrades
-				}
-				ts = tt
-			}
 		}
 	}
-
-	err = e.AddTradesToBuffer(resp...)
-	if err != nil {
+	if err := e.AddTradesToBuffer(resp...); err != nil {
 		return nil, err
 	}
 	resp = trade.FilterTradesByTime(resp, timestampStart, timestampEnd)
-
 	sort.Sort(trade.ByDate(resp))
 	return resp, nil
 }
 
 // SubmitOrder submits a new order
 func (e *Exchange) SubmitOrder(ctx context.Context, s *order.Submit) (*order.SubmitResponse, error) {
-	if err := s.Validate(e.GetTradingRequirements()); err != nil {
+	if s == nil || *s == (order.Submit{}) {
+		return nil, common.ErrEmptyParams
+	}
```

Comment:
Any reason why this was changed?:


```suggestion
	if err := s.Validate(e.GetTradingRequirements()); err != nil {
		return nil, err
	}
```

Reply:
`s.Validate` is the way to go, panic on nil isn't accepted for a wrapper function like this. If `Validate` has unnecessary checks that can be handled in its own PR and the places where those checks are necessary can be added as `requirements protocol.TradingRequirements, opt ...validate.Checker` per exchange

While I've definitely softened on panics, I don't really see this part as good enough reason to just not call `Validate` which has the check. Can remove the nil check from `Validate` in its own PR like I said, but `Validate` should still be called

Reply:
Thanks for reverting @samuael!

Reply:
I'm happy. Thanks @gloriousCode 
Resolving.